src/back/spv/writer.rs

```diff
@@ -103,30 +96,36 @@ impl Writer {
         }
     }
 
-    fn get_type_id(
-        &mut self,
-        arena: &crate::Arena<crate::Type>,
-        handle: crate::Handle<crate::Type>,
-    ) -> Word {
-        match self.lookup_type.lookup_id(handle) {
-            Some(word) => word,
-            None => {
-                let (instruction, id) = self.parse_type_declaration(arena, handle);
-                instruction.to_words(&mut self.logical_layout.declarations);
-                id
+    fn get_type_id(&mut self, arena: &crate::Arena<crate::Type>, lookup_ty: LookupType) -> Word {
+        let mut sink: Vec<(LookupType, Word)> = vec![];
+
+        let id = match self.lookup_type.entry(lookup_ty) {
+            Entry::Occupied(e) => *e.get(),
+            Entry::Vacant(e) => {
+                let id = match lookup_ty {
+                    LookupType::Handle(handle) => self.write_type_declaration_arena(arena, handle, &mut sink),
+                    LookupType::Local(local_ty) => self.write_type_declaration_local(arena, local_ty),
+                };
+                *e.insert(id)
             }
+        };
```

Comment:
I am getting the "cannot borrow `*self` as mutable more than once at a time" over here. I understand why, but I am not sure yet how to resolve this yet. 

The functions have to be mutable, even if we extract the part `let id = self.generate_id()` from these functions and use it here.

Reply:
hmm, right. I suppose we need to refactor `write_type_declaration_arena` to not work with `self`. Or at least, that's one way to address this...
if it's too difficult, we can go back to your original code, which I think was correct, and I'll look at it later to try to improve

Reply:
Yes I would prefer that. This means, the rest of the comments are also resolved?

Reply:
sure! Are we all set for landing then?